stetho/src/main/java/com/facebook/stetho/inspector/elements/Document.java

```diff
@@ -0,0 +1,602 @@
+
+package com.facebook.stetho.inspector.elements;
+
+import android.os.SystemClock;
+import com.facebook.stetho.common.Accumulator;
+import com.facebook.stetho.common.ArrayListAccumulator;
+import com.facebook.stetho.common.LogUtil;
+import com.facebook.stetho.common.ThreadBound;
+import com.facebook.stetho.common.UncheckedCallable;
+import com.facebook.stetho.inspector.helper.ObjectIdMapper;
+import com.facebook.stetho.inspector.helper.ThreadBoundProxy;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Queue;
+import java.util.regex.Pattern;
+
+public final class Document extends ThreadBoundProxy {
+  private final DocumentProvider.Factory mFactory;
+
+  private DocumentProvider mDocumentProvider;
+  private ShadowDocument mShadowDOM;
+  private ObjectIdMapper mObjectIdMapper;
+
+  private UpdateListenerCollection mUpdateListeners;
+
+  private ChildEventingList mCachedChildEventingList;
+  private ArrayListAccumulator<Object> mCachedChildrenAccumulator;
+
+  private int mReferenceCounter;
+
+  public Document(DocumentProvider.Factory factory) {
+    super(factory);
+
+    mFactory = factory;
+    mObjectIdMapper = new DOMObjectIdMapper();
+    mReferenceCounter = 0;
+    mUpdateListeners = new UpdateListenerCollection();
+  }
+
+  public synchronized void addRef() {
+    if (mReferenceCounter++ == 0) {
+      mDocumentProvider = mFactory.create();
+    }
+  }
+
+  public synchronized void release() {
+    if (mReferenceCounter > 0) {
+      if (--mReferenceCounter == 0) {
+        cleanUp();
+      }
+    }
+  }
+
+  private void cleanUp() {
+    mDocumentProvider.postAndWait(new Runnable() {
+      @Override
+      public void run() {
+        mDocumentProvider.setListener(null);
+        mShadowDOM = null;
+        mObjectIdMapper.clear();
+        mDocumentProvider.dispose();
+        mUpdateListeners.clear();
+        mDocumentProvider = null;
+      }
+    });
+  }
+
+  public void addUpdateListener(UpdateListener updateListener) {
+    mUpdateListeners.add(updateListener);
+  }
+
+  public void removeUpdateListener(UpdateListener updateListener) {
+    mUpdateListeners.remove(updateListener);
+  }
+
+  public NodeDescriptor getNodeDescriptor(Object element) {
+    verifyThreadAccess();
+    return mDocumentProvider.getNodeDescriptor(element);
+  }
+
+  public void highlightElement(Object element, int color) {
+    verifyThreadAccess();
+    mDocumentProvider.highlightElement(element, color);
+  }
+
+  public void hideHighlight() {
+    verifyThreadAccess();
+    mDocumentProvider.hideHighlight();
+  }
+
+  public void setInspectModeEnabled(boolean enabled) {
+    verifyThreadAccess();
+    mDocumentProvider.setInspectModeEnabled(enabled);
+  }
+
+  public int getNodeIdForElement(Object element) {
+    verifyThreadAccess();
+    return mObjectIdMapper.getIdForObject(element);
+  }
+
+  public Object getElementForNodeId(int id) {
```

Comment:
`ObjectIdMapper.getObjectForId` can return null and is marked as `@Nullable`, so we should mark this method as `@Nullable` too